plutus-wallet-api/ledger/Ledger/Typed/Tx.hs

```diff
@@ -0,0 +1,419 @@
+    tyTxForge        :: !Value.Value,
+    tyTxValidRange   :: !SlotRange
+    }
+
+baseTx :: TypedTx '[] '[]
+baseTx = TypedTx {
+    tyTxTypedTxIns = HNilF,
+    tyTxPubKeyTxIns = [],
+    tyTxTypedTxOuts = HNilF,
+    tyTxPubKeyTxOuts = [],
+    tyTxForge = Value.zero,
+    tyTxValidRange = Interval.always
+    }
+
+-- | Adds a 'TypedScriptTxOut' to a 'TypedTx'.
+--
+-- Adding a new typed transaction output is only safe if there are no typed transaction
+-- inputs yet. Otherwise those inputs would need to change to take the new data script as
+-- an argument.
+addTypedTxOut
+    :: forall outs newOut
+    . TypedScriptTxOut newOut
+    -> TypedTx '[] outs
+    -> TypedTx '[] (newOut ': outs)
+-- We're changing the type so we can't use record update syntax :'(
+addTypedTxOut out TypedTx {
+    tyTxTypedTxOuts,
+    tyTxPubKeyTxOuts,
+    tyTxTypedTxIns=HNilF,
+    tyTxPubKeyTxIns,
+    tyTxForge,
+    tyTxValidRange } = TypedTx {
+      tyTxTypedTxOuts=HConsF out tyTxTypedTxOuts,
+      tyTxPubKeyTxOuts,
+      tyTxTypedTxIns=HNilF,
+      tyTxPubKeyTxIns,
+      tyTxForge,
+      tyTxValidRange }
+
+-- | Adds a 'TypedScriptTxIn' to a 'TypedTx'.
+addTypedTxIn
+    :: forall ins outs newIn
+    . TypedScriptTxIn outs newIn
+    -> TypedTx ins outs
+    -> TypedTx (newIn ': ins) outs
+-- We're changing the type so we can't use record update syntax :'(
+addTypedTxIn inn TypedTx {
+    tyTxTypedTxOuts,
+    tyTxPubKeyTxOuts,
+    tyTxTypedTxIns,
+    tyTxPubKeyTxIns,
+    tyTxForge,
+    tyTxValidRange } = TypedTx {
+      tyTxTypedTxOuts,
+      tyTxPubKeyTxOuts,
+      tyTxTypedTxIns=HConsF inn tyTxTypedTxIns,
+      tyTxPubKeyTxIns,
+      tyTxForge,
+      tyTxValidRange }
+
+-- | A wrapper around a 'TypedTx' that hides the input list type as an existential parameter.
+-- This allows us to perform some operations more easily by not caring about the input connection
+-- types, see 'addSomeTypedTxIn' particularly.
+data TypedTxSomeIns (outs :: [Type]) = forall ins . TypedTxSomeIns (TypedTx ins outs)
+
+-- | Add a 'TypedScriptTxIn' to a 'TypedTxSomeIns'. Note that we do not have to track
+-- the input connection types explicitly.
+addSomeTypedTxIn
+    :: forall (outs :: [Type]) (newIn :: *)
+    . TypedScriptTxIn outs newIn
+    -> TypedTxSomeIns outs
+    -> TypedTxSomeIns outs
+addSomeTypedTxIn inn (TypedTxSomeIns tx) = TypedTxSomeIns $ addTypedTxIn inn tx
+
+-- | Adds many homogeneous 'TypedScriptTxIn' to a 'TypedTx'.
+addManyTypedTxIns
+    :: forall (ins :: [Type]) (outs :: [Type]) (newIn :: Type)
+    . [TypedScriptTxIn outs newIn]
+    -> TypedTx ins outs
+    -> TypedTxSomeIns outs
+addManyTypedTxIns ins tx = foldl' (\someTx inn -> addSomeTypedTxIn inn someTx) (TypedTxSomeIns tx) ins
+
+-- I'm sure I should be able to write these generically, but I think I would need quantified constraints?
+-- i.e. forall a . Coercible (f a) o
+hfCoerceIns :: forall d (ins :: [Type]) . HListF (TypedScriptTxIn d) ins -> [TxIn]
```

Comment:
I think you can abstract the recursion at least. By making a combinator that receives `forall a. f a -> b` and maps that function over `HListF` collecting results in a list. Then you'll just supply the same `coerce` to instantiate the combinator at different types. Or/and yes, I think quantified constraints should work.

Reply:
Perhaps something like the [`Forall`](https://hackage.haskell.org/package/row-types-0.3.0.0/docs/Data-Row-Internal.html#t:Forall) class from `row-types`?